File: 061/cpp/main.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == nullptr || k == 0) {
            return head;
        }
        
        int n = 0;
        ListNode *first = head;
        for (; n < k && first->next != nullptr; n++, first = first->next) ;

        ListNode *second = head;
        if (first->next == nullptr) {
            std::cout << n << std::endl;
            n++;
            k = n - 1 - k % n;
            for (int i = 0; i < k; i++, second = second->next) ;
        } else {
            for (; first->next != nullptr; first = first->next, second = second->next) ;
        }

        if (first == second) {
            return head;
        }
        ListNode *root = second->next;
        second->next = nullptr;
        first->next = head;
        return root;
    }
};

File: 061/rust/main.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn rotate_right(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
        let mut temp = head.as_deref();
        let mut v = Vec::new();
        while temp.is_some() {
            let t = temp.unwrap();
            v.push(t.val);
            temp = t.next.as_deref();
        }

        if v.is_empty() {
            return None;
        }
        let k = v.len() - 1 - (k as usize) % v.len();
        let mut root = None;
        for i in 0..v.len() {
            let j = if i > k { v.len() - (i - k) } else { k - i };
            let mut temp = ListNode::new(v[j]);
            temp.next = root;
            root = Some(Box::new(temp));
        }
        return root;
    }
}
